sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4

```diff
@@ -1502,8 +1502,11 @@ version
     ;
 
 operatorPipeRightSide
-    : selectClause
-    | whereClause
+    : selectClause windowClause?
+    // Note that the WINDOW clause is not allowed in the WHERE pipe operator, but we add it here in
+    // the grammar simply for purposes of catching this invalid syntax and throwing a specific
+    // dedicated error message.
+    | whereClause windowClause?
```

Comment:
Why we only allow optional window clause after SELECT and WHERE?

Reply:
According to @dtenedor Danial's offline discussion with Jeff Shute from the original research paper [1], we shall add the WINDOW clause as an optional operator at the end of existing operators that support expressions such as SELECT and WHERE. 

> We find that having |> WINDOW as a separate pipe operator breaks the composability of the operators. Therefore, we should actually add the WINDOW clause as optional at the end of existing operators that support expressions, like |> SELECT and |> WHERE.

Reply:
This is correct, this changes maintains the independent composability of the operators, and clarifies to which step the WINDOW clause applies.